### DISCUSSION TITLE
How can I use GNNExplainer and visualize the subgraph?

### DISCUSSION THREAD

**Discussion by masterqi912** [pyg-team/pytorch_geometric#7716] [Q&A] [answered]
Hi,

I tried the tutorials for XAI, however, although PyG has been renewed, the tutorials of GNN has not been renewed yet. The example offered in this blog post has been deleted. https://medium.com/@pytorch_geometric/graph-machine-learning-explainability-with-pyg-ff13cffc23c2

This is the code I used to construct the GNN on MUTAG dataset. https://colab.research.google.com/drive/1fLJbFPz0yMCQg81DdCP5I8jXw9LoggKO?usp=sharing

Then I applied code comes from the blog post:
```python
from torch_geometric.explain import Explainer, GNNExplainer
# Initialize explainer
explainer = Explainer(
    model=model,
    algorithm=GNNExplainer(epochs=200),
    explanation_type='model',
    node_mask_type='attributes',
    edge_mask_type='object',
    model_config=dict(
        mode='multiclass_classification',
        task_level='node',
        return_type='log_probs',
    ),
)
```

It couldn't offer explanation for me. Could you please offer examples of GNNExplainer for graph classification problem? Besides, could you please let me know how to visualize the subgraph of explanation?

Thank you!

-David

**Comment by wsad1** [comment, marked as answer]
Check out the example [here](https://github.com/pyg-team/pytorch_geometric/blob/9bc70173159b44502d6908c1563cf8a72cac5cdb/examples/explain/gnn_explainer.py#L63). You need to call `visualize_graph` to visualize the explanations.

**Reply by masterqi912** [reply]
Hi,

Thanks for your reply! I have successfully trained GCN on MUTAG dataset, according to the website you offered I have modified the codes of explainer as follows:

from tqdm import tqdm
from torch_geometric.explain import Explainer, GNNExplainer
model.eval()

# Initialize explainer
explainer_pyg = Explainer(
    model=model,
    algorithm=GNNExplainer(epochs=200),
    explanation_type='model',
    node_mask_type='attributes',
    edge_mask_type='object',
    model_config=dict(
        mode='binary_classification',
        task_level='node',
        return_type='raw',
    ),
)

for data in tqdm(train_loader):
  explanation_pyg = explainer_pyg(
      x=data.x,
      edge_index=data.edge_index,
      batch=data.batch,
      target=data.y
  )

However, I don't know how to call it on my test dataset. The examples and tutorials are more related to node classification, could you please offer help with graph classification task?

Thank you!

-David

**Reply by masterqi912** [reply]
I just solved this problem using following code:
```python
model.eval()

# The index of the graph in the dataset you want to explain
graph_index = 165

# Get the specific graph data
data = dataset[graph_index]

# Generate the explanation for the graph
explanation_pyg = explainer_pyg(
    x=data.x,
    edge_index=data.edge_index,
    batch=data.batch,
    target=data.y
)
```

However, the subgraph generated is hard to interprete since it is MUTAG dataset, a collection of chemical compounds. I would like to polish the result.
<img width="236" alt="image" src="https://github.com/pyg-team/pytorch_geometric/assets/129177649/3b28513d-3006-4f16-897a-89e7ee621512">

Besides, I am wondering whether GNNExplainer offers other results besides visulization of explanation?

Thanks!

-David

**Reply by wsad1** [reply]
You could pass just one batch to the explainer to get more interpretable explanations. 
Other than visualizing the fimportance graph. You could visualize the feature importance, or simply view the `available_explanations`. See [here](https://github.com/pyg-team/pytorch_geometric/blob/9bc70173159b44502d6908c1563cf8a72cac5cdb/examples/explain/gnn_explainer.py#L56)

**Reply by masterqi912** [reply]
Hi,

Thank you for your reply! When I had a look at the examples you offered: [GNNExplainer on BAShapes](https://github.com/pyg-team/pytorch_geometric/blob/9bc70173159b44502d6908c1563cf8a72cac5cdb/examples/explain/gnn_explainer_ba_shapes.py), I have questions.

1. Why do we need to use node index?
<img width="559" alt="image" src="https://github.com/pyg-team/pytorch_geometric/assets/129177649/8058663b-7f6f-4d87-8191-406783e367e9">
I thought it is a graph classification problem, so the code should be similar to MUTAG graph classification problem.

```python
### Initialize explainer
explainer_pyg = Explainer(
    model=model,
    algorithm=GNNExplainer(epochs=200),
    explanation_type='model',
    node_mask_type='attributes',
    edge_mask_type='object',
    model_config=dict(
        mode='binary_classification',
        task_level='node',
        return_type='raw',
    ),
)

explanation = explainer_pyg(
    x=data.x,
    edge_index=data.edge_index,
    batch=data.batch,
    target=data.y
)
```

Then I could use explanation.visualize_graph to plot the subgraph.

2.  Why do we need to train the explainer?
<img width="570" alt="image" src="https://github.com/pyg-team/pytorch_geometric/assets/129177649/a1b864a0-c34d-4a1f-805f-4a9978ca8029">

Since GNNExplainer is an instance-level explainer, we could directly generate explanation for a graph without training process. Maybe according to this step you would like to calculate AUC?

Thanks！

-David

**Reply by rusty1s** [reply]
1. This is a node-level classification task, and so we iterate over all node indices here for which we want to obtain an explanation.
2. `GNNExplainer` will train each instance separately. Maybe I misunderstood your question though.

**Reply by masterqi912** [reply]
Hi,

Thank you for your reply! I was confused since this step is not necessary for graph classification task, but I understand this step now.

-David

**Reply by sumone-compbio** [reply]
@masteryqq hi, did you figure out how to represent molecules with GNNExplainer results? There is an option to highlight molecules' substructures in rdkit (https://www.rdkit.org/docs/GettingStartedInPython.html) but you have to provide the string for the substructure. Highlighting nodes/edges in gradient based on importance scores that we get from GNNExplainer would be nice.
